woocommerce/woocommerce-android
# Feat/products 2706 add product action

## RESOURCES
[The Figma file](https://www.figma.com/file/QAkdm5HZdLlG0b8vUkv1pX/Products---Add-%26-Edit-Products?node-id=460%3A1760&viewport=-1704%2C1577%2C0.5334523320198059) and [the initial issue](https://github.com/woocommerce/woocommerce-android/issues/2707)

## CHANGES / DESCRIPTION
The first set change(s) for the **Add new Products** feature.

This PR will focus on:
- adding a `M4` release feature flag
- displaying the FAB button in the `ProductListFragment`
- showing the bottom sheet fragment for `ProductTypes`

Took some inspiration from [this PR](https://github.com/woocommerce/woocommerce-android/pull/2712) and refactored things a bit. 

**EDIT:**
From @0nko valid PR comments, I made some refactor work on this PR. 

To really make the most out of the work @anitaa1990 had done in the #2712 PR, I introduced a new argument to the navigation actions for the product type bottom sheet. Using the already familiar `navArgs`, it will let the VM/Fragment know if the inflate type is `isAddProduct` (which defaults to `false` to keep the original bottom sheet behaviour from the `ProductDetails` screen).

Using this approach and, as suggested, the `loadProductTypes` will then let the VM know of the type of list **builder** to use. Inside the VM, using the `savedState.navArgs()`, we can also get the Fragment's arguments and handle the items selections depending on the (once again) argument flag of `isAddProduct`.

One extra thing to be aware is that, in order to have the bottom sheet fragment accessible from both the `ProductList` and the `ProductDetails` screen, I had to move the `productTypesBottomSheetFragment` to a new navigation file called `nav_graph_product_type_bottom_sheet` (which is where the arguments are defined for the Fragment).

To close the loop, the `nav_graph_main` now has a `action_global_productTypeBottomSheetFragment` and the `nav_graph_products` has its own `action_productDetailFragment_to_productTypesBottomSheetFragment`. 

This approach saves us from having 2 very similar Fragments and VM's, reusing most of the work 👌 
Let me know your thoughts on this work so far 👍 

**P.S -** I merged the latest `develop` from the upstream, so some other bits got added to the PR for some reason. I hope this is alright.

## TESTING
- Click on the PRODUCT bottom nav bar button
    - Notice the loading screen not showing the FAB button
    - Notice the FAB button appearing after the screen loads
    - Click on the FAB button
    - Notice the bottom sheet for `Product Add types` appearing with the product list showing behind
 - Scenario 1
   - Click in the outside area of the bottom sheet fragment
   - Notice the bottom sheet fragment disappearing
 - Scenario 2
   - Click in any item of the list
   - Notice the bottom sheet fragment disappearing

## SCREENSHOTS
Product List screenshots
-

<img src="https://user-images.githubusercontent.com/7058393/90012922-33c85f80-dc9c-11ea-96d2-4fab0fe5be00.png" width="200"/> <img src="https://user-images.githubusercontent.com/7058393/90012962-417de500-dc9c-11ea-87c0-849d932af2ce.png" width="200"/>

Product List interaction
-
<img src="https://user-images.githubusercontent.com/7058393/90018135-61190b80-dca4-11ea-9ba9-e577cdd3928c.gif" width="200"/>


Product Detail screenshot
-
<img src="https://user-images.githubusercontent.com/7058393/90013329-f87a6080-dc9c-11ea-857a-3974fcc425d4.png" width="200"/>
